### ISSUE TITLE
Expected quantities vs Measured quantities

### ISSUE THREAD

**Issue by pricelessbrewing** [beerjson/beerjson#56]
Obviously beersmith supports both, but IIRC @rocketman768 brewtarget only supports estimated values. I know there's empty fields in the brewday print outs, but I don't think brewtarget has any way to input measurements? Brewersfriend does not support measured values. Brewtoad does not support measured values. Beersmith does support it, and does export some of the properties in the recipe.xml file, notably abv vs est abv, and efficiency vs actual efficiency, batch size vs display batch size, etc etc. 

As it stands I believe all quantities are currently the expected/estimated values? 

If we are going to support both software estimated, and real life measured values, how will we do so? 

A) Store the respective values in every property? My preference, as it keeps things following the current organization method rather than having to reproduce each schema for a measured copy.

IE

```
"boil_size": {
          "units": "gal",
          "volume": 5.5
},
```

Becomes

```
"boil_size": {
          "units": "gal",
          "volume": 
          {
          "expected": 5.5,
          "measured": 5.25
          }
},
```

or 

B) Have a separate file for all measured values. IE Estimated recipe.json and Measured recipe.json

**Comment by rocketman768** [comment]
Brewtarget has a "Brew Notes" feature which allows you to record actual measurements. @mikfire wrote that, so he may have some comments.

Option A has the disadvantage of only allowing a single measurement I think. What if I brew the recipe 10 times? If you suggest an array for measured values, it gets a little funky since it requires corresponding different values by index. I'd think it's cleaner to have a record separate from a recipe with measured values for each "brewing."

**Comment by krutilin** [comment]
Agree with @rocketman768. Logging records should be separate from the recipe. Maybe we should have some brewlog structure which will store brewing session data.

**Comment by mikfire** [comment]
I am always disappointed that so few people know of the brewnotes.

tl;dr You will want a structure something like this:
```json
[
   { 
      "brewer": "mikfire",
      "assistant": "his evil twin",
      "brewdate": 2018-03-04,
      "abv": {
         "expected": 5.4,
         "projected": 5.8,
         "measured": 5.7
     }
]
```

It does have to be it's own data structure. You will want what I would call an array of hashes, aka ```[{}]```. 

The estimated v measured is interesting. I think I finally used three: calculated, projected and measured. Consider what happens when your recipe expects 75% efficiency, but you get 80%. That would not change the calculated ABV since that is based on the recipe and the expected efficiency. It would change the projected ABV, since that is based on what you actually brewed. And of course, there is the measured ABV after fermentation.

Another issue I had was dealing with changes to the original recipe. Imagine I brew a recipe three times and decide it needs a higher ABV, so I add another kilo of base malt. In my opinion, that should not change any of the information in the three batches I've already brewed. I ended up storing the original fields that I needed from the recipe into the brewnote.

I would like to see things like "brewer", "assistant", all of the primary/secondary/tertiary temperatures/dates/times removed from the recipe specification and put into this. Those possibly change with every brew day and should not be single values in the recipe object itself. Same for things like taste notes.

You may want to look at the BrewNotes objects in brewtarget and its database. I think it might provide some valuable input into this. I won't claim it's the best solution, but it has worked quite well for me.

**Comment by rmsrosa** [comment]
I use BeerTools Pro, which does not have this feature, but I always though it would be nice to have it. 

But I worry that the more features we add to the beerjson standard, the harder it becomes for the software developers to adhere to it. For this reason, I agree that the recipe should have just one field (which would essentially correspond to the "expected"/"planned" one, but not necessarily), and the brew note to be a separate, non-required record.

I am not so sure about including all three "expected", "projected", and "measured" fields, though. If I understood it correctly, the idea is that "projected" is useful midway through the process. I can see the value of having this in a brewing software, but it is less important in a recipe-exchange standard. We could make it simpler and possibly better-adopted by having the "array of hashes" just with the fields "planned" and "measured" (or "expected" or "estimated" instead of "planned", although I like "planned" better).

**Comment by pricelessbrewing** [comment]
@mikfire 

>brewnotes

I don't see that in the program. I see extras tab, with taste notes, and notes text fields, and the fermentation stages, but I don't see for example Measured OG, measured FG, measured volumes etc. 

>Expected vs projected vs measured

I'm not sure all three are necessary. I see quite a bit of confusion already between beersmiths expected vs estimated fields, project isn't really any better. 

Expected/estimated/projected should be the main recipes base value, and measured would correspond to a specific brew "session" so to speak on the software side, for the schema each measured value set would correspond to a new array object.

**Comment by mikfire** [comment]
Yeah. That's been the problem. Right click on a recipe and select Brew
it.
Mik

On Mon, Mar 5, 2018, at 18:24, Priceless Brewing wrote:
> @mikfire[1]


>> brewnotes


> I don't see that in the program. I see extras tab, with taste notes,
> and notes text fields, and the fermentation stages, but I don't see
> for example Measured OG, measured FG, measured volumes etc.>> Expected vs projected vs measured


> I'm not sure all three are necessary. I see quite a bit of confusion
> already between beersmiths expected vs estimated fields, project isn't
> really any better.> Expected/estimated/projected should be the main recipes base value,
> and measured would correspond to a specific brew "session" so to speak
> on the software side, for the schema each measured value set would
> correspond to a new array object.> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[2], or mute the thread[3].> 

Links:

  1. https://github.com/mikfire
  2. https://github.com/beerjson/beerjson/issues/56#issuecomment-370604081

**Comment by pricelessbrewing** [comment]
Bringing this up again since it's been a couple months. 

Really this depends on the scope of beerjson, from my current point of view it's a matter of 

1) "is brew session data relevant to other people?" IMO no, they just care about the end recipe when sharing data between people. 

2) In a certain aspect, a brew session is just a duplicate of the recipe with slightly different values. If a recipe fits the beerjson schema, then so will the brew session copy of it, they will simply have slightly different values. Both will fit the recipe schema.

**Comment by mikfire** [comment]
As a counter argument to your first point, if I want to port my recipes
from brewtarget to brewtoad, then I care very, very much about that
history. It would be nice to have some sort of interchange format for
that information.
I do find the second point an interesting observation. 

There may be a path using this idea. What if we added an optional
"childOf" attribute that simply stored the ID of a "parent" recipe. A
brew session would then simply copy the recipe being brewed (the
"parent"), and populate that attribute with the appropriate ID in the
new copy (the "child").

Mik

On Sun, Aug 5, 2018, at 21:34, Mark Price wrote:
> Bringing this up again since it's been a couple months.


> Really this depends on the scope of beerjson, from my current point of
> view it's a matter of


>  1. "is brew session data relevant to other people?" IMO no, they just
>     care about the end recipe when sharing data between people.


>  2. In a certain aspect, a brew session is just a duplicate of the
>     recipe with slightly different values. If a recipe fits the
>     beerjson schema, then so will the brew session copy of it, they
>     will simply have slightly different values. Both will fit the
>     recipe schema.> — You are receiving this because you were mentioned. Reply to this
> email directly, view it on GitHub[1], or mute the thread[2].

Links:

  1. https://github.com/beerjson/beerjson/issues/56#issuecomment-410565608

**Comment by pricelessbrewing** [comment]
Sure, but is that necessary for the recipe schema standard or should that be internal to the brewing software? For recipe exchange between users or software shouldn't the two simply be standard recipe files?

**Comment by pricelessbrewing** [comment]
Fair enough. It would be nice to standardize some inter-software standards like inventory, brew sessions, and ingredient database, I just don't know if that's part of the goal of this project. @beerjson/beerjson-wg  ?

I think having the brew sessions as daughter copies would be simplest, as it wouldn't require a separate schema, and there's already a notes section i believe? So it would be as simple as that, in which case it seems like it would be pretty minimal to support in that way.

**Comment by pricelessbrewing** [comment]
@beerjson/beerjson-wg 

Would it be possible to have a measurement.json that basically followed beer.json, but did not have any required fields? So it would support all properties that would be available in the original recipe, but by default would be blank until the software/user said there was a deviation from the original recipe expectations?

**Comment by RossMerr** [comment]
@pricelessbrewing  I already encountered that problem myself in my development but relay as I see it everything under the RecipeType is agnostic to your equipment so you can scale it etc, so by that logic you can't have measured values and everything has to be calculated. 

As I see it we have a following scenarios,

1. RecipeType need to stay agnostic to your equipment for the primary objective of sharing a recipe for someone else to brew.
2. We want the measured values recorded to help give as much information as we can for someone else brewing the recipe.
3. We want to refine a recipe from the measured values, but is a refined recipe still the same original recipe? And then once it has been refined would those original measured values still relevant? maybe to yourself but not likely for others,
4. We want to record the measured values from multiple batches for measuring beer consistency (maybe include data from things like Tilt).

I'd say point 1 is already fine with the current schema and we should not modify the RecipeType and all its children.
Point 2 and 4 are a very similar problem, brew a batch as best as we can to the recipe just point 4 wants a collection.
Point 3 is a more a abstract problem to defining as its still in the design phase of the recipe and so should be for the software/user to define when its a new recipe.

With that all said I think we need like a BatchType collection on the root of the schema for each brew of the recipe to store all the measured values plus also have a relationship to the EquipmentType so you know what it was brewed on and the RecipeType.

**Comment by RossMerr** [comment]
@pricelessbrewing I put together a branch for what I was thinking, it's a bit rough so any feedback is welcome 

https://github.com/RossMerr/beerjson/blob/56_quantities/json/batch.json

**Comment by pricelessbrewing** [comment]
@RossMerr @krutilin 

I'm not experienced enough to know if this is even possible, but it seems like we should be able to have a way to do something like a batch.json or measurements.json that's a duplicate of beer.json (so it supports all possible properties), but all of the required fields are not required. 

Maybe we need to add a script to duplicate all of the files in /json/ and make a copy in a subfolder /measurements/ or just rename like measurements_boil.json is a duplicate of boil.json but boil_time isn't required? 

Otherwise we could leave it up to the software and make the software keep track of a recipe vs a measurement file. There's really no difference as all measured properties are also recipe properties. So the measured values would be the  

For example, if I had a recipe where the preboil volume was supposed to be 5.5 gallons, but my measured value was 6.1 gallons, the two files would either be identical apart from that value, or the measurements.json would be empty apart from the preboil size.

**Comment by RossMerr** [comment]
@pricelessbrewing At the moment I favour your idea of leaving it to the software/developer. 👍 

We could re-look at this problem once we get more feedback from real word implementation and then extend the schema if necessary, nothing wrong with that.

My main reason for that is I'm personally I'm still not convinced it'll relay works with just copying the current schema and removing the required fields, Yeah it's got a lot of similarities and it's could definitely be automated but it's also missing stuff like, I want to record multiple temperature, pH and gravity readings as I'm fermenting that's not currently possible. I do think we would end up with lots of unnecessary properties like name, and description and yes they are optional and would probably be ignored but then why would we want them in the schema, doesn't make things very clean for anyone.
